tools/rpmuncompress.c

```diff
@@ -151,26 +152,39 @@ static char *doUntar(const char *fn)
     if ((at = getArchiver(fn)) == NULL)
 	goto exit;
 
-    if (dstpath) {
-	    int sr = singleRoot(fn);
+    int needtar = (at->extractable == 0);
 
-	    /* the trick is simple, if the archive has multiple entries,
-	     * just extract it into the specified destination path, otherwise
-	     * strip the first path entry and extract in the destination path
-	     */
-	    rasprintf(&mkdir, "mkdir '%s' ; ", dstpath);
+    if (dstpath) {
+	char * sr = singleRoot(fn);
+
+	/* the trick is simple, if the archive has multiple entries,
+	 * just extract it into the specified destination path, otherwise
+	 * strip the first path entry and extract in the destination path
+	 */
+	rasprintf(&mkdir, "mkdir '%s' ; ", dstpath);
+	if (needtar)
 	    rasprintf(&stripcd, " -C '%s' %s", dstpath, sr ? "--strip-components=1" : "");
+	else {
+	    char * moveup;
+	    if (sr)
+		rasprintf(&moveup, " ; tmp=`mktemp -u -p'%s'` ; mv %s/%s $tmp ; (shopt -s dotglob; mv $tmp/* %s) ; rmdir $tmp ", dstpath, dstpath, sr, dstpath);
```

Comment:
We know the name of the directory the archive will create, we shouldn't need any globs here? Make a temporary directory, use that as the target directory to extract to (since all the relevant commands seem to support such an option), then rename to the intended location and remove the tmpdir? Assuming that's workable, we could use the same thing for tar and avoid using separate logic for it and the GNU specific --strip-components option. (Updating the tar to use the newly added field was implied in my comment in the ticket, sorry it wasn't entirely clear - always assuming it's possible, BUT our assumptions didn't exactly hold here)

This is all pretty hard to follow though, if we need a shell script then maybe we should write one, or at least make it move obvious by eg placing it in a multiline string section that *looks* like a shell script to begin with :sweat_smile:

Reply:
Nah, we need to move the files out of the top level dir. We could extract into the tmpdir - which would be cleaner and saver, but there is no way around moving the files. We can't rename the top level dir to the destination as we can't assume the destination dir is empty or this is the only extraction we do during build.

Reply:
Re-formatted the script part a bit to make it easier to read.

Reply:
Hmm, I didn't think of multiple unpacks to the same -C directory. I don't know how common a scenario that would be but it's a fair point. But, if that's a goal then lets test for it too, because right now it fails:

> $ mkdir /tmp/aabb
$ ./tools/rpmuncompress -x -C /tmp/aabb ../tests/data/SOURCES/test-1.2.3.zip 
mkdir: cannot create directory ‘/tmp/aabb’: File exists

If we care about collisions then we should indeed unpack to a temporary directory initially, it's possible there's a directory that conflicts whereas its contents don't. Should also avoid one of the mv's?

Reply:
OK, added -p to the mkdir. Also reworked the whole logic to create the tmpdir first and extract into that.